objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+        // will always result in a valid field element.
+        asset[0] = Felt::new((asset[0].as_int() << 1) >> 1);
+
+        // set the 32 least significant bits of the 3rd element to the non-fungible asset tag
+        let temp = (asset[2].as_int() >> 32) << 32;
+        asset[2] = Felt::new(temp | AccountId::NON_FUNGIBLE_FAUCET_TAG as u64);
+
+        Ok(Self(asset))
+    }
+
+    // HELPER FUNCTIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an error if:
+    /// - The first bit of the asset is not ZERO.
+    /// - The 32 least significant bits of the 3rd element are not set to 2^31.
+    fn validate(&self) -> Result<(), AssetError> {
+        let first_bit = self.0[0].as_int() >> 63;
+        if first_bit != 0 {
+            return Err(AssetError::non_fungible_asset_invalid_first_bit());
+        }
+
+        let tag = self.0[2].as_int() as u32;
+        if tag != AccountId::NON_FUNGIBLE_FAUCET_TAG {
+            return Err(AssetError::NonFungibleAssetInvalidTag(tag));
+        }
+
+        Ok(())
+    }
+}
+
+impl Deref for NonFungibleAsset {
+    type Target = Word;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<NonFungibleAsset> for Word {
+    fn from(asset: NonFungibleAsset) -> Self {
+        asset.0
+    }
+}
+
+impl From<NonFungibleAsset> for [u8; 32] {
+    fn from(asset: NonFungibleAsset) -> Self {
+        let mut result = [0_u8; 32];
+        result[..8].copy_from_slice(&asset.0[0].as_int().to_le_bytes());
+        result[8..16].copy_from_slice(&asset.0[1].as_int().to_le_bytes());
+        result[16..24].copy_from_slice(&asset.0[2].as_int().to_le_bytes());
+        result[24..].copy_from_slice(&asset.0[3].as_int().to_le_bytes());
+        result
+    }
+}
+
+impl From<NonFungibleAsset> for Asset {
+    fn from(asset: NonFungibleAsset) -> Self {
+        Asset::NonFungible(asset)
+    }
+}
+
+impl TryFrom<Word> for NonFungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: Word) -> Result<Self, Self::Error> {
+        let asset = Self(value);
+        asset.validate()?;
+        Ok(asset)
+    }
+}
+
+impl TryFrom<[u8; 32]> for NonFungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: [u8; 32]) -> Result<Self, Self::Error> {
+        let word = parse_word(value)?;
+        Self::try_from(word)
+    }
+}
+
+impl fmt::Display for NonFungibleAsset {
+    fn fmt(&self, _f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        todo!()
+    }
+}
+
+// NON-FUNGIBLE ASSET DETAILS
+// ================================================================================================
+
+/// Details about a non-fungible asset.
+///
+/// Unlike [NonFungibleAsset] struct, this struct contains full details of a non-fungible asset.
+pub struct NonFungibleAssetDetails {
+    faucet_id: AccountId,
+    asset_data: String,
```

Comment:
Any reason why we want to use a string here instead of `Vec<u8>`? If we want to enforce some display pattern, we could alternatively apply that only to the `core::fmt::Display` implementation and allow the user to freely store his data as he wants.

Reply:
Good question! Actually, there is no good reason for this to be a string, but also, I don't want to leave the task of serializing data to the user. Ideally, it would be something like JSON, but not a the string level. Basically, it would be good to enforce (even if loosely), the following properties:

- The data should be serializable into bytes.
- The data should be format-independent. For example, adding whitespace or order of fields should not change the hash.
- The data should be reducible to a display string which reflex internal content. For example, we could iterate over internal fields and show a value for each field.

One potential option is to use something like [Serde JSON](https://github.com/serde-rs/json), but I also think this is something that we need to consider more carefully. So, for now, I've changed the data type to `Vec<u8>` and we can come back to it later to figure out the proper solution.

Reply:
Added #4 to keep track of this.